### DISCUSSION TITLE
Alma 8 no Temperature

### DISCUSSION THREAD

**Discussion by asalois** [ClementTsang/bottom#1185] [Q&A] [answered]
Hello,

I built using 'cargo install --locked bottom'

When I launch btm it shows "no data" for the temperatures. I have lm_sensors installed. All btop, ytop and sensors report temperatures but not bottom why? Older Linux kernel 4.18? 

Thanks in Advance!

**Comment by ClementTsang** [comment]
Could you give more information on your platform? What version/distro of Linux (I'm assuming it's AlmaLinux from your title, but I need more details on which specific version)? What architecture?

**Reply by ClementTsang** [reply]
Also would be possibly really helpful if I could get a screenshot or something, just to see if anything else looks off.

**Reply by ClementTsang** [reply]
I think(?) I have an idea of why it might not be working btw - if you don't mind helping me troubleshoot and manually compiling, I can send you a branch with a potential bug fix.

**Reply by ClementTsang** [reply]
I can't test locally since it's tricky to set up temp sensors in a VM, but if it's not too much trouble, could you try the branch here: https://github.com/ClementTsang/bottom/pull/1186

Specifically:

```shell
git clone https://github.com/ClementTsang/bottom
git checkout linux_do_not_immediately_fail_temps
cargo run
```

and let me know if there are any temperatures displayed?

**Reply by asalois** [reply]
For more information, I am running Alma 8 on a HP 17 envy laptop via ssh. I have included screenshots for btop and sensors to show that other software is reporting temperatures correct. I have also included a screenshot from the standard build from 'cargo install --locked btottom"

![hp17-alma8](https://github.com/ClementTsang/bottom/assets/30240707/5faaa76b-f74d-4941-bae4-333ec59c5c39)
![normal-build](https://github.com/ClementTsang/bottom/assets/30240707/4e64f771-ca0e-4cdc-a430-812b9356e6c1)
![sensors](https://github.com/ClementTsang/bottom/assets/30240707/6f5fac36-a9b3-4bbd-ac7b-09eb6082059d)
![btop](https://github.com/ClementTsang/bottom/assets/30240707/e0258fde-4bbc-45a2-b400-d83c41fff0c0)

**Reply by asalois** [reply]
Here is the screenshot from the branch you specified. It shows temperatures but does not show per core temperatures like btop does. Having the package temperature is enough for me but wanted to point this out.
 
![branch_linux_do_not_immediately_fail_temps](https://github.com/ClementTsang/bottom/assets/30240707/683f914f-caee-4d61-9387-b632dc0a39a4)

**Reply by ClementTsang** [reply]
Hm, there in another directory than what we expect in the code.

What does `ls /sys/devices/platform/coretemp.*` return?

**Reply by ClementTsang** [reply]
Also thanks a bunch for verifying the first bug fix, I'll aim to publish a release with it and other bug fixes sometime this week.

**Reply by asalois** [reply]
Here is the output plus a tree. 

alex @ hp17➜  ~ ls -r /sys/devices/platform/coretemp.*
uevent  subsystem  power  modalias  hwmon  driver_override  driver
alex @ hp17➜  ~ tree /sys/devices/platform/coretemp.*
/sys/devices/platform/coretemp.0
├── driver -> ../../../bus/platform/drivers/coretemp
├── driver_override
├── hwmon
│   └── hwmon1
│       ├── device -> ../../../coretemp.0
│       ├── name
│       ├── power
│       │   ├── autosuspend_delay_ms
│       │   ├── control
│       │   ├── runtime_active_time
│       │   ├── runtime_status
│       │   └── runtime_suspended_time
│       ├── subsystem -> ../../../../../class/hwmon
│       ├── temp1_crit
│       ├── temp1_crit_alarm
│       ├── temp1_input
│       ├── temp1_label
│       ├── temp1_max
│       ├── temp2_crit
│       ├── temp2_crit_alarm
│       ├── temp2_input
│       ├── temp2_label
│       ├── temp2_max
│       ├── temp3_crit
│       ├── temp3_crit_alarm
│       ├── temp3_input
│       ├── temp3_label
│       ├── temp3_max
│       ├── temp4_crit
│       ├── temp4_crit_alarm
│       ├── temp4_input
│       ├── temp4_label
│       ├── temp4_max
│       ├── temp5_crit
│       ├── temp5_crit_alarm
│       ├── temp5_input
│       ├── temp5_label
│       ├── temp5_max
│       └── uevent
├── modalias
├── power
│   ├── autosuspend_delay_ms
│   ├── control
│   ├── runtime_active_time
│   ├── runtime_status
│   └── runtime_suspended_time
├── subsystem -> ../../../bus/platform
└── uevent

8 directories, 40 files

**Reply by ClementTsang** [reply]
Yep, looks like I need to query this path as well. I'll get to it later today. Thanks for all the cooperation!

**Reply by ClementTsang** [reply]
Made a PR (https://github.com/ClementTsang/bottom/pull/1188), if you could try the `linux_temps_check_sys_devices` branch could you let me know if it works for you? Seems to work fine for me on my Intel laptop but would be good to verify it works for your use case.

**Reply by asalois** [reply]
Assuming I actually switched branches correctly, there is no difference on the temperature display. I even tried to scroll through the temperatures and there are only 4. 
![Screenshot 2023-06-08 071250](https://github.com/ClementTsang/bottom/assets/30240707/c4e39cfb-8fb3-4cd0-8372-5434b82180c6)
![Screenshot 2023-06-08 071205](https://github.com/ClementTsang/bottom/assets/30240707/54b0749c-65ee-4e02-90a2-71d893b8fe37)

**Reply by ClementTsang** [reply]
Hmm... weird. I'll check again when I get home, might have changed something with the latest commit.

**Reply by ClementTsang** [reply]
Hm. I tried it at home on my Intel laptop, and it seems to work fine for me in terms of extracting the CPU temps from the directory I was asking about above.

Just to be sure, on that branch (which seems right), can you do:

```sh
git pull
cargo run
```

and select the temperature widget, then press `e` to make sure they're not just hidden?

EDIT: for example, this is what I see if I disable hwmon sensors on my laptop:
![image](https://github.com/ClementTsang/bottom/assets/34804052/6502f001-d1b6-4843-b141-7eab389c2e26)

**Reply by asalois** [reply]
What do you mean disable hwmon?

To get the following screenshots I did a new git clone and git checkout. 

![Screenshot 2023-06-09 064822](https://github.com/ClementTsang/bottom/assets/30240707/cd15463b-b490-497b-879a-40467ece24ab)
![Screenshot 2023-06-09 064951](https://github.com/ClementTsang/bottom/assets/30240707/0d0487c7-3f2f-4fa9-9696-cba90bb32597)

**Reply by ClementTsang** [reply]
Sorry, that just meant I had gone into the code and dropped any data gathered using a certain method, so what was shown is only the type of data that should be obtained on your machine as well.

**Reply by ClementTsang** [reply]
Alright, did some changes - if you could pull and build, let me know if anything has changed.

If it's still not working, could you please run:

```bash
cargo run --features logging
```

And attach the generated `debug.log` file in the same folder? 

Also thank you so much for helping me throughout figuring out this issue!

**Reply by asalois** [reply]
Hello, here is what things look like now. I have included the debug.log for you. As seen in the screenshots there is more sensor readings but we some with no labels now.

[debug.log](https://github.com/ClementTsang/bottom/files/11712667/debug.log)
![Screenshot 2023-06-10 084533](https://github.com/ClementTsang/bottom/assets/30240707/69a2f793-f63f-43dc-83f5-401e40f15bbb)
![Screenshot 2023-06-10 084516](https://github.com/ClementTsang/bottom/assets/30240707/be9a9e39-998f-4e43-907b-1b72d7560445)

**Reply by ClementTsang** [reply]
Cool, seems like I fixed one bug but possibly introduced another in the process.

**Reply by ClementTsang** [reply]
Made more changes to make the sensor name parsing less short-circuit-y.

Same deal, try running it, and if it really doesn't work, then run `cargo run --features logging` for like 10 seconds and upload the `debug.log` file. Thanks!

**Reply by asalois** [reply]
Looks like it did not fix the issue. 
![bottom-no-temps](https://github.com/ClementTsang/bottom/assets/30240707/2dfadd29-66a6-44bb-b376-d453013e42b0)
[debug.log](https://github.com/ClementTsang/bottom/files/11713734/debug.log)

**Reply by ClementTsang** [reply]
Sorry, should have mentioned to please pull - from the debug log, seems like you're running the previous version.

**Reply by asalois** [reply]
I did but something looked fishy, so I removed the repo and cloned again followed by 'git checkout linux_temps_check_sys_devices'
Does this help?

EDIT I broke the log link before
[debug.log](https://github.com/ClementTsang/bottom/files/11713740/debug.log)
![bottom-no-temps-01](https://github.com/ClementTsang/bottom/assets/30240707/48793a64-d6cb-46b5-90cf-39eff83f8f9a)

**Reply by ClementTsang** [reply]
Yeah, that helps - though this is *weird*, it seems like some of your sensors are reporting... empty names for me if I parse through `/sys/class/hwmon`, but they work fine if I use `/sys/class/thermal`...

Why they worked before was because I would immediately stop reading from `/sys/class/hwmon` if I saw something wrong in *any* of the name entries and fall back to `/sys/class/thermal`, but this meant I was dropping the `coretemp` entries.

**Reply by ClementTsang** [reply, marked as answer]
Alright, seems like there was one case where I didn't use the thermal label which *should* be fixed now - if possible, could you try pulling and testing again? Thanks!

**Reply by asalois** [reply]
You got it! Things are looking good! Here is the debug.log if you want to look at it. 

![bottom-works](https://github.com/ClementTsang/bottom/assets/30240707/d94eb237-737a-4977-847e-fb4938523e81)
[debug.log](https://github.com/ClementTsang/bottom/files/11714977/debug.log)

**Reply by ClementTsang** [reply]
Nice to see it working! I'll clean some stuff up and probably publish a release sometime today or tomorrow.

Thanks so much for being patient and working along with me throughout, this wouldn't be possible otherwise.
